### ISSUE TITLE
Mongoose 3.8.26 gives bizarre results when used with Q promises

### ISSUE THREAD

**Issue by alavers** [Automattic/mongoose#2864]
I use [Q.ninvoke](https://github.com/kriskowal/q) with mongoose and I've started seeing bizarre behavior after upgrading from 3.8.25 to 3.8.26. The following code snippet will reproduce the issue:

```
var Q = require('q');
var mongoose = require('mongoose');
var Schema = mongoose.Schema;

var ArticleSchema = new Schema({
    title: String,
    tags: [String]
});

var Article = mongoose.model('Article', ArticleSchema);

return Q.ninvoke(Article, 'update', {
    title: 'Hello'
}, {
    $push: {
        tags: {
            $each: ['nifty', 'awesome']
        }
    }
}).then(function(res) {
    return Q.ninvoke(Article, 'findOne', {
        title: 'Hello'
    });
}).then(function(article) {
    console.log(article);
});
```

Supposing we with a document `{title:'Hello', tags:[]}`, after running the above code the tags some how get double inserted to give this:  `{title:'Hello', tags:['nifty', 'awesome', 'nifty', 'awesome']}`

This code works as expected in 3.8.25.
Also, this issue does not repro if I invoke `mongoose.update` directly, without using Q.

**Comment by wavded** [comment]
Experiencing issues with Bluebird promises as well, reverted to 3.8.25 and it worked again.

**Comment by vkarpov15** [comment]
Well this is a wonky bug, courtesy of https://github.com/aheckmann/mquery/commit/3cca43758f1e9905f79ce368dd88240f30665311 . mquery >= 1.1 has a `.then()` function, so Q thinks that the return value of `Article.findOne()` is a promise. That function should have been introduced in 2.0 to mark it as backwards breaking in hindsight. This is an unintentional side effect of the fix for #1934, which bumped mquery version for 3.8.26. As you probably noticed, the below works:

```javascript
return Article.update({
    title: 'Hello'
}, {
    $push: {
        tags: {
            $each: ['nifty', 'awesome']
        }
    }
}).then(function(res) {
    return Article.findOne({
        title: 'Hello'
    });
}).then(function(article) {
    console.log(article);
});
```

All solutions to this are roughly equally unpleasing. The easiest solution will probably be just deleting `mquery.prototype.then` in mongoose, but that's a bit of a dirty hack. Will try and see what happens.
